Date Picker Core Component

jakubcolony opened the pull request:
I closed my previous DatePicker [pull request](https://github.com/JoinColony/colonyDapp/pull/3851) as I was made aware it should have originated from a different branch. I also connected the DatePicker to the streaming form.

## Description

This PR adds a new component, `DatePicker` that will allow users to pick date and, optionally, time. It's based on [React Datepicker](https://reactdatepicker.com/) library.

"Plain" DatePicker:
<img width="691" alt="image" src="https://user-images.githubusercontent.com/112586815/191586153-46727a3d-f89b-4437-a555-a938b0c61be1.png">

DatePicker with time select:
<img width="697" alt="image" src="https://user-images.githubusercontent.com/112586815/191586258-e9e3fbdb-bf6b-4110-91d5-aa2ed31c2ae7.png">

DatePicker with options:
<img width="698" alt="image" src="https://user-images.githubusercontent.com/112586815/191586364-578b91c6-7080-4338-b51c-cfc5217864f7.png">

**New stuff** ✨

 * `DatePicker` component
 * `date-fns` package for formatting Date objects
 * `@floating-ui` package - the designs required a time dropdown to be nested inside the date dropdown and the Popper library used by both `react-datepicker` and our internal components (`Popover`) cannot handle these very well
 * `patch-package` - the library above uses TypeScript 4+ features while the dapp runs on Typescript 3.x. This package allowed me to apply a custom "patch" to make it compatible with dapp's TS version. I spoke to @chinins about this and she confirmed the cdapp will use a newer version of TS so this is only a temporary solution.

**Changes** 🏗

**Deletions** ⚰️
* Deleted the old `DatePicker` component which wasn't used anywhere in the dapp.

## TODO

- [x] Basic DatePicker component
- [x] DatePicker with time selection
- [x] DatePicker within select (https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=1195%3A241884)

Resolves #3834

jakubcolony commented:
@willm30's comment on my previous PR:

>@jakubcolony Good work Jakub! I will take a closer look at the code, but for now my first thought is that this should be based off https://github.com/JoinColony/colonyDapp/pull/3833 (streaming payment), as it does not belong in the "advanced" expenditure type. Then you can connect it to "Starts" (and "Ends" too or leave that for a follow up pr if it adds too much additional size to this one).

Thank you for spotting this!

>Also a few notes on the styles:

> * in [the design](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=5755%3A331599) the border is not rounded and the border color is var(--grey-blue-2);
> * the scrollbar on time picker drodown should be rounded and styled per design
> * time dropdown border is var(--temp-grey-blue-8) [in figma](https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=6364%3A334590)

All should be fixed now. I checked with Arren and it's okay to leave the system's default scrollbar

>But to reiterate, looks like really solid work. Good job!

Thanks! 👍

willm30 commented on src/modules/core/components/Fields/DatePicker/DatePicker.tsx:
```diff
@@ -0,0 +1,252 @@
+
+export interface DatePickerOption {
+  label: string | MessageDescriptor;
+  value: any;
```
How come value is `any` here instead of `string`?

jakubcolony replied:
That's an interesting one as I didn't want to restrict the type of the option's value. Many HTML elements (button, input) accept a `value` attribute which is of type `string | number | string[] | undefined` so I suppose I'll change my typing accordingly

willm30 commented on src/modules/core/components/Fields/DatePicker/DatePicker.tsx:
```diff
@@ -0,0 +1,252 @@
+      onClick={onClick}
+      ref={ref}
+    >
+      {currentOption && currentOption.hideDatePicker ? labelText : value}
```
Could read `currentOption?.hideDatePicker ?...`

willm30 commented on src/modules/dashboard/components/ExpenditurePage/ExpenditureSettings/ExpenditureStreamingForm.tsx:
```diff
@@ -39,14 +40,16 @@ export const MSG = defineMessages({
+    hideDatePicker: true,
   },
   {
     label: MSG.fixedTime,
```
Add `hideDatePicker: false` to this one for consistency? And then remove as optional property from `DatePickerOption`? Purely a matter of personal preference.

willm30 reviewed (changes requested):
My first pass impression is that this is a solid component! Left a few small notes on the code but overall looks fabulous. 

Main question is with this UTC / local business. 

Time is being displayed as UTC, but stored in the form state as local. Why this difference? Should we not just store it as UTC if we're also displaying it as UTC?

![dates](https://user-images.githubusercontent.com/64402732/191736214-41845908-8bca-4424-a74b-5a8d6a26891f.png)

I also notice in Figma the date string displayed explicitly includes "UTC" at the end. Not sure if that's been omitted intentionally. 

Separately, the component goes off screen for longer months in the End Date section:

![offscreen](https://user-images.githubusercontent.com/64402732/191736807-6c8fb8c1-3d8d-4aab-b44c-b32933501067.png)

And currently the default state looks like the selected state. Default in figma is just the color highlight, without the background.

Figma:
![def](https://user-images.githubusercontent.com/64402732/191737250-3a979b5d-9a9f-4a54-bbce-2c07e5d31b23.png)

Branch:
![branch](https://user-images.githubusercontent.com/64402732/191737393-ad25846a-7cfc-406b-9b95-9f56727837b0.png)

But good job!

willm30 commented:
Oh, and btw. I don't think you invited `colony/frontend` to review this pr? I think I got here via the old one. Also just noting the branch naming convetion generally includes the issue number. And finally, do you know if this component is intended to supersede the existing date picker we have (`components/DatePicker`)? Or we'll hang on to both?

jakubcolony commented on src/modules/core/components/Fields/DatePicker/DatePicker.tsx:
```diff
@@ -0,0 +1,252 @@
+      updatedDate.setUTCMinutes(UTCDate.getMinutes());
+
+      helpers.setValue({
+        ...field.value,
```
I'm overriding date but want to persist `option` property

jakubcolony commented:
>Main question is with this UTC / local business.

>Time is being displayed as UTC, but stored in the form state as local. Why this difference? Should we not just store it as UTC if we're also displaying it as UTC?

Actually, all Date objects in JavaScript store a single moment in time in UTC but "the basic methods to fetch the date and time or its components all work in the local (i.e. host system) time zone and offset" ([mdn](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date)). Adding that to the limitations of `date-fns` library, I need to convert the correct UTC date into an offset date which doesn't represent the correct moment in time, but makes the "local component" of Date equal to the correct UTC date (eg. 10 AM UTC becomes 10 AM BST/CEST/whichever timezone you're in). 
I hope this explanation is clear but ping me on Discord and I'll be happy to go over it in more detail

>And currently the default state looks like the selected state. Default in figma is just the color highlight, without the background.

I was under impression this isn't marking the default but today's date:
<img width="355" alt="image" src="https://user-images.githubusercontent.com/112586815/191843194-82b7e83d-f9a7-42e4-85bc-ce63d6127b86.png">
I think "selected" and "default" are interchangeable in this case as there must be some "selected" date by default.
@arrenv are you able to confirm this?

>And finally, do you know if this component is intended to supersede the existing date picker we have (components/DatePicker)? Or we'll hang on to both?

I don't know but the existing one doesn't seem to be used anywhere. @arrenv should I get rid of the old one?

willm30 commented on src/modules/pages/components/ExpenditurePage/constants.ts:
```diff
@@ -54,6 +54,13 @@ export const initialValues = {
     ],
   },
+  streaming: {
+    startDate: { date: new Date() },
```
With respect to setting the UTC time, could you not do it here? Then you wouldn't need to do those conversions onChange. 

Eg:

``` js

const nowUTC = () => {
  const date = new Date();
  const utc = new Date(
    date.getUTCFullYear(),
    date.getUTCMonth(),
    date.getUTCDate(),
    date.getUTCHours(),
    date.getUTCMinutes(),
  );
  return utc;
};
{
  startDate: { date: nowUTC() }
...
}

```

And remove lines 98-102 and 122-132 in `DatePicker`, and just have the component always reading from the form state (`value.date`).

jakubcolony replied:
Since the "local UTC time" represents an incorrect moment in time, I think it makes more sense to do all the hacky conversion in the DatePicker component and only ever expose the correct Date object. I checked with Arren and we're not planning support for timezones other than UTC so this solution seemed like the most sensible one

willm30 commented on src/modules/core/components/Fields/DatePicker/DatePicker.tsx:
```diff
@@ -0,0 +1,261 @@
+    <div>
+      <ReactDatePicker
+        selected={UTCDate}
+        onChange={handleDateChange}
```
Should do  `setTouched(true)` on change also?

willm30 commented on src/modules/core/components/Fields/DatePicker/TimePicker.tsx:
```diff
@@ -0,0 +1,119 @@
+  );
+
+  const timeOptions = useMemo(() => {
+    return [...new Array((24 * 60) / 30)].map((_, idx) => {
```
Small one. `24 * 2` makes more sense to me intuitively. "24 hours in a day, times 2 to get half hours".

jakubcolony replied:
I've only just realised there's a bug in this line. The value `30` should be `timeInterval` instead

willm30 reviewed (commented):
Added a few comments. Also, when using the "Ends" date picker, the month of October is still going off screen for me.

jakubcolony commented:
@arrenv as discussed I changed the picker to work in local timezone instead of UTC and also added a tooltip that shows you the selected date in UTC. Regarding the timezone abbreviations, it seems pretty difficult to reliably get the abbreviation of the user's timezone so I opted for the default "GMT-x" notation:
<img width="391" alt="image" src="https://user-images.githubusercontent.com/112586815/192054762-dd41808d-f504-45c8-b02e-4d6bbf466f83.png">

willm30 commented on src/modules/core/components/Fields/DatePicker/TimePicker.tsx:
```diff
@@ -0,0 +1,120 @@
+
+import Icon from '~core/Icon';
+import styles from './DatePicker.css';
+import { DEFAULT_TIME_FORMAT } from './constants';
```
I'm always the one who forgets to check this, but styles come last in the imports list ⬆️🆙

willm30 commented on src/modules/core/components/Fields/DatePicker/DatePicker.tsx:
```diff
@@ -0,0 +1,275 @@
+  MONTH_NAMES,
+} from './constants';
+import styles from './DatePicker.css';
+import { Tooltip } from '~core/Popover';
```
Same as below (styles last)

willm30 reviewed (approved):
Nice work Jakub! Looks good to me. 🙌🙌

arrenv commented on src/modules/core/components/Fields/DatePicker/constants.ts:
```diff
@@ -0,0 +1,18 @@
+export const MONTH_NAMES = [
```
These should probably use our language system.

arrenv reviewed (changes requested):
@jakubcolony Looks great, nice work.

A couple of comments and requests.

1. I think we should get rid of the `GMT +X`. It should simplify the appearance, and with the very nice tooltip, it provides enough explanation for the timezones.

2. The alignment of the month and the arrows is off.

![32786407-d9f808a75974763b7169e304794c5fed](https://user-images.githubusercontent.com/33682027/192289403-ba012b25-47df-4f32-836c-3381c9766614.png)

3. However, with that, there is an oversite from the product side. I think we really need to include the year in the date picker.

This has been updated here - https://www.figma.com/file/dCtvv76S8mk5HNApjwmBto/Expenditures?node-id=6364%3A334590

4. We have six caret arrows and two down ones. Do we need to introduce another?

Julianahlk reviewed (approved):
Great work Jakub! The date picker is working smoothly 🙌
<img width="409" alt="Screen Shot 2022-09-27 at 12 12 40 AM" src="https://user-images.githubusercontent.com/71563622/192390509-1ad63079-9cd5-44a5-9e73-44fc634828c8.png">
<img width="731" alt="Screen Shot 2022-09-27 at 12 18 47 AM" src="https://user-images.githubusercontent.com/71563622/192390600-175ea56b-1f18-4dee-bb48-7fa0541a5e69.png">
---
I do have a question regarding old dates, should the user be able to set a start date in the past? And would it make more sense to grey out any date in the past to make it clearer that is also an old date?

arrenv commented:
> I do have a question regarding old dates, should a user be able to set a start date in the past? And would it make more sense to grey out any date in the past to make it clearer that the selected date is also an old date?

Good pickup @Julianahlk, for inspiration, @thomassdk solved this here fairly recently - https://github.com/JoinColony/coinMachine/pull/196

jakubcolony commented:
>The alignment of the month and the arrows is off.

Can you check now? They're aligned on my machine:
<img width="478" alt="image" src="https://user-images.githubusercontent.com/112586815/192627430-2b92d922-0f5a-4562-8aea-7cb5c346cfbe.png">

I added an optional `minDate` prop (similarly to Thomas's solution) which disables the days and times before the specified min date. 
@arrenv which colour should I use for past dates?

arrenv commented:
> > The alignment of the month and the arrows is off.
> 
> Can you check now? They're aligned on my machine: <img alt="image" width="478" src="https://user-images.githubusercontent.com/112586815/192627430-2b92d922-0f5a-4562-8aea-7cb5c346cfbe.png">
> 
> I added an optional `minDate` prop (similarly to Thomas's solution) which disables the days and times before the specified min date. @arrenv which colour should I use for past dates?

Hey @jakubcolony, you can use this color `--temp-grey-6`.

jakubcolony commented:
@arrenv that has now been updated:
<img width="335" alt="image" src="https://user-images.githubusercontent.com/112586815/193134117-e693d9a4-1b27-4872-ac26-341236035c48.png">

arrenv commented on src/modules/dashboard/components/ExpenditurePage/ExpenditureSettings/ExpenditureStreamingForm.tsx:
```diff
@@ -39,18 +40,21 @@ export const MSG = defineMessages({
   {
     label: MSG.limitIsReached,
     value: ExpenditureEndDateTypes.LimitIsReached,
+    hideDatePicker: true,
```
Wouldn't it be better to have this as toggle on, i.e. showDatePicker: true, then default, should be not having to include it.
